std/typecons.d

```diff
@@ -5893,3 +5893,135 @@ public:
     auto value = cast(int)flags_A;
     assert(value == Enum.A);
 }
+
+/**
+   Type constructor for final variables.
+
+   A final variable is "head-const"; it cannot be directly
+   mutated or rebound, but references reached through the variable
+   are typed with their original mutability.
+ */
+struct Final(T)
+    if(!is(T == const) && !is(T == immutable))
+{
+    private:
+    T data;
+
+    public:
+    /// $(D Final) subtypes $(D T) as an rvalue.
+    inout(T) Final_get() inout pure nothrow @nogc @safe
+    {
+        return data;
+    }
+
+    alias Final_get this; /// Ditto
+
+    ///
+    this(T data)
+    {
+        this.data = data;
+    }
+
+    /// Ditto
+    this(Args...)(auto ref Args args)
+        if(__traits(compiles, T(args)))
+    {
+        this.data = data;
```

Comment:
As far as I see, your code is the same as 
    
    this.data = this.data;

which is practically a no-op. Your unittest does not check if any field actually gets assigned. Of course it compiles, but it does not do anything meaningful.